File: Insulator/app/src/main/java/com/bendywalker/insulator/Calculator.java
package com.bendywalker.insulator;

import java.math.RoundingMode;
import java.text.DecimalFormat;

public class Calculator {
    private static final int MGDL_CONVERSION_VALUE = 18;

    private double carbohydrateFactor;
    private double correctiveFactor;
    private double desiredBloodGlucose;
    private double currentBloodGlucose;
    private double carbohydratesInMeal;
    private double totalDailyDose;
    private BloodGlucoseUnit bloodGlucoseUnit;

    public Calculator(double carbohydrateFactor, double correctiveFactor, double desiredBloodGlucose, double currentBloodGlucose, double carbohydratesInMeal, BloodGlucoseUnit bloodGlucoseUnit) {
        this.carbohydrateFactor = carbohydrateFactor;
        this.correctiveFactor = correctiveFactor;
        this.desiredBloodGlucose = desiredBloodGlucose;
        this.currentBloodGlucose = currentBloodGlucose;
        this.carbohydratesInMeal = carbohydratesInMeal;
        this.bloodGlucoseUnit = bloodGlucoseUnit;
    }

    public Calculator(double totalDailyDose, BloodGlucoseUnit bloodGlucoseUnit) {
        this.totalDailyDose = totalDailyDose;
        this.bloodGlucoseUnit = bloodGlucoseUnit;
    }

    private double convertBloodGlucose(double bloodGlucose) {
        switch (bloodGlucoseUnit) {
            case mmol:
                return bloodGlucose;
            case mgdl:
                return bloodGlucose / MGDL_CONVERSION_VALUE;
            default:
                return bloodGlucose;
        }
    }

    public double getCarbohydrateDose() {
        double carbohydrateDose = 0.0;

        if (carbohydrateFactor != 0) {
            carbohydrateDose = carbohydratesInMeal / carbohydrateFactor;
        }
        carbohydrateDose = round(carbohydrateDose);
        return carbohydrateDose;
    }

    public double getCorrectiveDose() {
        double correctiveDose = 0.0;

        if (currentBloodGlucose != 0) {
            correctiveDose = (convertBloodGlucose(currentBloodGlucose) - convertBloodGlucose(desiredBloodGlucose)) / correctiveFactor;
        }
        correctiveDose = round(correctiveDose);
        return correctiveDose;
    }

    public double getTotalDose() {
        double totalDose = getCarbohydrateDose() + getCorrectiveDose();

        if (totalDose < 0) {
            totalDose = 0.0;
        }

        return totalDose;
    }

    public double getCarbohydrateFactor() {
        carbohydrateFactor = 500 / totalDailyDose;
        carbohydrateFactor = round(carbohydrateFactor);
        return carbohydrateFactor;
    }

    public double getCorrectiveFactor() {
        double correctiveFactor = 0.0;

        switch (bloodGlucoseUnit) {
            case mmol:
                correctiveFactor = 100 / totalDailyDose;
                break;
            case mgdl:
                correctiveFactor = (100 / totalDailyDose) * MGDL_CONVERSION_VALUE;
        }

        correctiveFactor = round(correctiveFactor);
        return correctiveFactor;
    }

    public static double round(Double value) {
        DecimalFormat decimalFormat = new DecimalFormat("#0.0");
        decimalFormat.setRoundingMode(RoundingMode.HALF_UP);
        return Double.valueOf(decimalFormat.format(value));
    }
}

File: Insulator/app/src/main/java/com/bendywalker/insulator/Calculator.kt
package com.bendywalker.insulator

import com.bendywalker.insulator.extension.round

/**
 * Created by Ben David Walker (bendywalker) on 06/01/2017.
 */

/**
 * Calculates different dosage values given a set of parameters.
 *
 * @param currentBloodGlucose The user's current blood glucose level.
 * @param carbohydratesInMeal Grams of carbohydrates in the user's meal.
 * @param carbohydrateFactor The user's carbohydrate factor.
 * @param correctiveFactor The user's corrective factor.
 * @param desiredBloodGlucose The blood glucose level the user wishes to attain.
 * @param bloodGlucoseUnit The user's preferred blood glucose unit (mmol/L or mg/dL).
 */
class Calculator(currentBloodGlucose: Double = 0.0, var carbohydratesInMeal: Double = 0.0,
                 private val carbohydrateFactor: Double, correctiveFactor: Double,
                 desiredBloodGlucose: Double, private val bloodGlucoseUnit: BloodGlucoseUnit) {

    /**
     * Constructs a [Calculator] used to calculate different dosage values. The majority of parameters required for
     * calculations are retrieved from [SharedPreferences] via [PersistedValues].
     *
     * @param currentBloodGlucose The user's current blood glucose level
     * @param carbohydratesInMeal Grams of carbohydrates in the user's meal
     * @param persistedValues An instance of [PersistedValues] used to read calculation parameters from [SharedPreferences]
     */
    constructor(currentBloodGlucose: Double = 0.0, carbohydratesInMeal: Double = 0.0, persistedValues: PersistedValues) :
            this(currentBloodGlucose, carbohydratesInMeal, persistedValues.carbohydrateFactor, persistedValues.correctiveFactor, persistedValues.desiredBloodGlucose, persistedValues.bloodGlucoseUnit)

    private val correctiveFactor = correctiveFactor
        get() = convertBloodGlucose(field)

    private val desiredBloodGlucose = desiredBloodGlucose
        get() = convertBloodGlucose(field)

    var currentBloodGlucose: Double = currentBloodGlucose
        get() = convertBloodGlucose(field)

    val carbohydrateDose: Double
        get() {
            if (carbohydrateFactor != 0.0) {
                val carbohydrateDose = carbohydratesInMeal / carbohydrateFactor
                return carbohydrateDose.round(roundingPrecision)
            } else return 0.0
        }

    val correctiveDose: Double
        get() {
            if (correctiveFactor != 0.0) {
                val correctiveDose = (currentBloodGlucose - desiredBloodGlucose) / correctiveFactor
                return correctiveDose.round(roundingPrecision)
            } else return 0.0
        }

    val totalDose: Double
        get() {
            val totalDose = carbohydrateDose + correctiveDose
            return if (totalDose > 0) totalDose.round(roundingPrecision) else 0.0
        }

    private fun convertBloodGlucose(bloodGlucose: Double): Double {
        when (bloodGlucoseUnit) {
            BloodGlucoseUnit.MMOL -> return bloodGlucose
            BloodGlucoseUnit.MGDL -> return (bloodGlucose / mgdlConversionValue).round(roundingPrecision)
            else -> return bloodGlucose
        }
    }

    companion object {
        private val mgdlConversionValue = 18
        private val roundingPrecision = 1

        /**
         * Given a *total daily dose*, calculates a user's carbohydrate factor.
         *
         * @param totalDailyDose The sum of all insulin taken daily by a user.
         * @return Calculated carbohydrate factor.
         */
        fun carbohydrateFactor(totalDailyDose: Double): Double {
            val carbohydrateFactor = 500 / totalDailyDose
            return carbohydrateFactor.round(roundingPrecision)
        }

        /**
         * Given a *total daily dose*, calculates a user's corrective factor.
         *
         * @param totalDailyDose The sum of all insulin taken daily by a user.
         * @param bloodGlucoseUnit A user's preferred blood glucose unit.
         * @return Calculated corrective factor.
         */
        fun correctiveFactor(totalDailyDose: Double, bloodGlucoseUnit: BloodGlucoseUnit): Double {
            val correctiveFactor: Double

            when (bloodGlucoseUnit) {
                BloodGlucoseUnit.MMOL -> correctiveFactor = 100 / totalDailyDose
                BloodGlucoseUnit.MGDL -> correctiveFactor = 100 / totalDailyDose * mgdlConversionValue
            }

            return correctiveFactor.round(roundingPrecision)
        }
    }
}
